File: Python/robot_test.py
import os
import cv2
import glob
import torch
import numpy as np

from modules import utils
from modules import models


# Define device
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# Root path of data
root = os.path.relpath('../dataset')

# Load stereo calibration parameters
Params = cv2.FileStorage(os.path.join(root,'StereoParams.xml'), cv2.FILE_STORAGE_READ)
K1 = Params.getNode('K1').mat()
dist1 = Params.getNode('dist1').mat()


# Create SuperPoint model
superpoint = models.SuperPointNet(3)
superpoint.load_state_dict(torch.load(os.path.join(root,'py_superpoint.pt'), map_location=device))

# Create EllipSegNet model
ellipsegnet = models.EllipSegNet(16, 1)
ellipsegnet.load_state_dict(torch.load(os.path.join(root,'py_ellipsegnet.pt'), map_location=device))

# Create MarkerPose model
markerpose = models.MarkerPose(superpoint, ellipsegnet, (320,240), 120, Params)
markerpose.to(device)
markerpose.eval()


# Set window name and size
cv2.namedWindow('Pose estimaion', cv2.WINDOW_NORMAL)
cv2.resizeWindow('Pose estimaion', 1792, 717)

# Read left and right images of the target
I1 = utils.natsort(glob.glob(os.path.join(root,'images','L','*')))
I2 = utils.natsort(glob.glob(os.path.join(root,'images','R','*')))


# Test
with torch.no_grad():
    for im1n, im2n in zip(I1,I2):
        im1 = cv2.imread(im1n)
        im2 = cv2.imread(im2n)

        # Pose estimation
        R, t = markerpose(im1, im2)

        # Visualize results
        utils.drawAxes(im1, K1, dist1, R, t)
        
        cv2.imshow('Pose estimaion',np.hstack([im1,im2]))
        if cv2.waitKey(500) == 27:
            break

    cv2.destroyAllWindows()

File: C++/robot_test.cpp
#include <iostream>
#include <filesystem>

#include <torch/torch.h>
#include <opencv2/opencv.hpp>

#include "src/models.hpp" // SuperPointNet, EllipSegNet, MarkerPose
#include "utils.hpp" // imlist, drawAxes

int main()
{
    // Defining device: CPU or GPU
    torch::DeviceType device = torch::cuda::cudnn_is_available() ? torch::kCUDA : torch::kCPU;

    // Root path of dataset
    std::filesystem::path root{"../../dataset"};

    // Load stereo calibration parameters
    cv::FileStorage Params;
    Params.open(root/"StereoParams.xml", cv::FileStorage::READ);
    cv::Mat K1, dist1;
    Params["K1"] >> K1;
    Params["dist1"] >> dist1;


    // Create SuperPoint model
    SuperPointNet superpoint(3);
    torch::load(superpoint, root/"cpp_superpoint.pt", device);

    // Create EllipSegNet model
    EllipSegNet ellipsegnet(1, 16);
    torch::load(ellipsegnet, root/"cpp_ellipsegnet.pt", device);

    // Create MarkerPose model
    MarkerPose markerpose(superpoint, ellipsegnet, cv::Size(320,240), 120, Params);
    markerpose->to(device);
    markerpose->eval();


    // Set window name and size
    cv::namedWindow("Pose estimaion", cv::WINDOW_NORMAL);
    cv::resizeWindow("Pose estimaion", 1792, 717);


    // List of images
    auto I1 = imlist(root/"images"/"L");
    auto I2 = imlist(root/"images"/"R");

    
    // Test
    torch::NoGradGuard no_grad;
    for (int i = 0; i < I1.size(); i++)
    {
        cv::Mat im1 = cv::imread(I1[i]);
        cv::Mat im2 = cv::imread(I2[i]);

        // Pose estimation
        auto [R, t] = markerpose(im1, im2);

        // Visualize results
        drawAxes(im1, K1, dist1, R, t);
        cv::Mat im;
        cv::hconcat(im1, im2, im);

        cv::imshow("Pose estimaion", im);
        if (cv::waitKey(500) == 27) break;
    }
}